assets/js/search.js

```diff
@@ -0,0 +1,218 @@
+
+  function detectSearchEntry() {
+    let entries = [];
+    if (window.libraryEntries) {
+
+      entries = window.libraryEntries;
+      return {
+        entries,
+        type: 'libraryEntries'
+      };
+
+    } else if (window.guideEntries) {
+
+      entries = window.guideEntries;
+      return {
+        entries,
+        type: 'guideEntries'
+      };
+    }
+  }
+
+  //Function to track Google analytics
+  // function trackGA {
+  //    // Track what users are searching for via Google Analytics events
+  //     //   ga('send', 'event', "iac-lib", "search", "query", text);
+  // }
+
+
+   /**
+   * A function to search the IaC Lib and Deployment guides. Can also be used for other pages that need it.
+   * To show/hide the proper elements based on the results. 
+   * @type {Function}
+   */
+  function filterData(event, type) {
+
+    let target = $(event.currentTarget);
+    let searchValue = target.val();
+
+    $('#no-matches').hide();
+
+    if (searchValue && searchValue.length > 0) {
+
+      let lowerText = searchValue.toLowerCase();
+      // if(typeof(searchQueries) == 'string'){
+      //   searchQueries.push(lowerText);
+      // }
+
+      let searchQueries = lowerText.split(" ");
+
+      if (searchEntry.type == 'libraryEntries') {
+        $('.table-clickable-row').hide();
+      } else if (searchEntry.type == 'guideEntries') {
+        $('.guide-card').hide() && $('.category-head').hide() && $('.categories ul').hide();
+      }
+
+      let entries = searchEntry.entries;
+
+      let matches = 0;
+      let submoduleMatches = 0;
+
+      for (let i = 0; i < entries.length; i++) {
+        let entry = entries[i];
+        let matchesAll = true;
+        for (let j = 0; j < searchQueries.length; j++) {
+          let searchQuery = searchQueries[j];
+
+          let searchContent;
+          if (searchEntry.type == 'libraryEntries') {
+            searchContent = entry.text;
+          } else if (searchEntry.type == 'guideEntries' && type == 'wordSearch') {
+            searchContent = entry.title + entry.category + entry.content + entry.tags;
+          } else if (searchEntry.type == 'guideEntries' && type == 'tagSearch') {
+            searchContent = entry.tags;
+          } else {
+            searchContent = entry.cloud;
+          }
+
+          if (searchContent.indexOf(searchQuery) < 0) {
+            matchesAll = false;
+            break;
+          }
+        }
+
+        if (matchesAll) {
+          $("#" + entry.id).show();
+          matches++;
+          (searchEntry.type == 'libraryEntries') ? submoduleMatches += entry.num_submodules: submoduleMatches = 0;
+        }
+
+      }
+      if (matches === 0) {
+        $('#no-matches').show();
+      }
+
+      showItemsCount(matches, submoduleMatches);
+    } else {
+      if (searchEntry.type == 'libraryEntries') {
+
+        showInitialItemsCount();
+        $('.table-clickable-row').show();
+
+      } else if (searchEntry.type = 'guideEntries') {
```

Comment:
Multiple problems here:

1. Single `=` is not a comparison. Not sure how this works at all? This and the comparison above should be `===`.
1. Hard-coded `type` value. We should have constants defined for these.
1. The `filterData` function says it can be used with other pages, but it's clearly hard-coded for just these two pages. This function should be refactored to actually be a generic search function... E.g., Pass in the data to search as an argument; pass in callbacks to handle how to render search results; etc. Otherwise, we are just shoving two implementations side by side into one function, which actually makes maintenance harder.

Reply:
> Not sure how this works at all? T
it "works" because it's an else if, so it already checked the first option and this is the only other option and an assignment always returns true 😅 So this reads as `if (searchEntry.type == 'libraryEntries') {} else {}`. Still bug in the code, but this is why it coincidentally works